File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    /*
        Great question on sliding window. 
        Approach: 
        1. Store the freq of each char so far in a map. 
        2. Store the maximum freq which is maximum number of times a char has occured so far. 
        3. If the window size - max_freq  are greater than given k, we need to reduce the window from the left. 
            3a. For that we need to remove the entry of left char from the map and increment left. 
        4. Store a potential window size by taking maximum from a global result. 
    
    */
    int characterReplacement(string s, int k) {
        int n = s.size();
        int left = 0;
        int right = 0;
        
        int max_freq = 0;
        
        unordered_map<char, int> mp;
        int res = 0;
        while (right < n) {
            mp[s[right]]++;
            max_freq = max(max_freq, mp[s[right]]);
            
            while ((right - left + 1) - max_freq > k) {
                mp[s[left]]--;
                left++;
            }
            res = max(res, right - left + 1);
            right++;
        }
        return res;
        
    }
};

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public void eraseFromMap(Map<Character, Integer> map, char ch) {
        if (map.get(ch) == 1) map.remove(ch);
        else map.put(ch, map.get(ch) - 1);
    }
    public int characterReplacement(String s, int k) {
        int n = s.length();
        int left = 0;
        int right = 0;
        int maxCount = 0;
        int res = 0;
        Map<Character, Integer> map = new HashMap<>();
        while (right < n) {
            char ch = s.charAt(right);
            map.put(ch, map.getOrDefault(ch, 0) + 1);
            maxCount = Math.max(maxCount, map.get(ch));

            while ((right - left + 1) - maxCount > k) {
                eraseFromMap(map, s.charAt(left));
                left++;
            }
            res = Math.max(res, right - left + 1);
            right++;
        }
        return res;
    }
}